File: test/frontend/submit.pl
#!/usr/bin/perl -w

use saliweb::Test;
use Test::More 'no_plan';
use Test::Exception;
use File::Temp;

BEGIN {
    use_ok('modloop');
}

my $t = new saliweb::Test('modloop');

# Check job submission

# Check get_submit_page
{
    my $self = $t->make_frontend();
    my $cgi = $self->cgi;

    my $tmpdir = File::Temp::tempdir(CLEANUP=>1);
    ok(chdir($tmpdir), "chdir into tempdir");

    ok(mkdir("incoming"), "mkdir incoming");
    ok(open(FH, "> test.pdb"), "Open test.pdb");
    print FH "REMARK\nATOM      2  CA  ALA     1      26.711  14.576   5.091\n";
    ok(close(FH), "Close test.pdb");
    open(FH, "test.pdb");

    throws_ok { $self->get_submit_page() }
              saliweb::frontend::InputValidationError,
              "no key";

    $cgi->param('pdb', \*FH);
    $cgi->param('name', 'test');
    $cgi->param('modkey', '***REMOVED***');
    $cgi->param('loops', '1::1::');
    my $ret = $self->get_submit_page();
    like($ret, qr/Job Submitted.*You can check on your job/ms,
         "submit page HTML");

    seek(FH, 0, 0);
    $cgi->param('email', 'test@test.com');
    $ret = $self->get_submit_page();
    like($ret, '/Job Submitted.*You will be notified.*' .
               'You can check on your job/ms',
         "submit page HTML");

    ok(unlink("incoming/input.pdb"), "remove input PDB file");
    ok(unlink("incoming/loops.tsv"), "remove loop selection");
}

# Test check_loop_selection
{
    modloop::check_loop_selection("anything");

    throws_ok { modloop::check_loop_selection("") }
              saliweb::frontend::InputValidationError,
              "check_loop_selection, empty loop";
    like($@, qr/No loop segments were specified/, 
         "                      exception message");
}

# Test check_pdb_name
{
    modloop::check_pdb_name("anything");

    throws_ok { modloop::check_pdb_name("") }
              saliweb::frontend::InputValidationError,
              "check_pdb_name, no file";
    like($@, qr/No coordinate file/,
         "                exception message");
}

# Test make_residue_id
{
    is(modloop::make_residue_id('A', '54'), '54:A', 'make_residue_id');
    is(modloop::make_residue_id(' b ', ' 35 '), '35:b',
       '                (whitespace)');
}

# Check parse_loop_selection
{
    my ($loops, $start_res, $start_id, $end_res, $end_id, $loop_data) =
        modloop::parse_loop_selection('1::5::16:A:30:A:');
    is(scalar @$loop_data, 8, 'parse_loop_selection loop_data');
    is($loop_data->[0], '1');
    is($loop_data->[1], ' ');
    is($loop_data->[2], '5');
    is($loop_data->[3], ' ');
    is($loop_data->[4], '16');
    is($loop_data->[5], 'A');
    is($loop_data->[6], '30');
    is($loop_data->[7], 'A');
    is(scalar @$start_res, 2, '                     start_res');
    is($start_res->[0], '1');
    is($start_res->[1], '16');
    is(scalar @$start_id, 2, '                     start_id');
    is($start_id->[0], ' ');
    is($start_id->[1], 'A');
    is(scalar @$end_res, 2, '                     end_res');
    is($end_res->[0], '5');
    is($end_res->[1], '30');
    is(scalar @$end_id, 2, '                     end_id');
    is($end_id->[0], ' ');
    is($end_id->[1], 'A');
    is($loops, 2, '                     loops');

    throws_ok { modloop::parse_loop_selection('1::5::16:A:30:') }
              "saliweb::frontend::InputValidationError",
              '                     wrong number of colons';
    like($@, qr/Syntax error in loop selection/,
         '                     error message');

    throws_ok { modloop::parse_loop_selection('1::5::16:A:20:B:') }
              "saliweb::frontend::InputValidationError",
              '                     loop that spans chains';
    like($@, qr/not selected properly.*16:A,.*20:B/ms,
         '                     error message');

    throws_ok { modloop::parse_loop_selection('1::50::16:A:20:B:') }
              "saliweb::frontend::InputValidationError",
              '                     loop that is too long';
    like($@, qr/not selected properly.*1: ,.*50: /ms,
         '                     error message');

    throws_ok { modloop::parse_loop_selection('10::5::16:A:20:B:') }
              "saliweb::frontend::InputValidationError",
              '                     loop that is negative length';
    like($@, qr/not selected properly.*10: ,.*5: /ms,
         '                     error message');

    throws_ok { modloop::parse_loop_selection('1::10::20:A:31:A:') }
              "saliweb::frontend::InputValidationError",
              '                     too many residues';
    like($@, qr/Too many loop residues have been selected.*22 > limit:20/,
         '                     error message');

    throws_ok { modloop::parse_loop_selection('') }
              "saliweb::frontend::InputValidationError",
              '                     no residues';
    like($@, qr/No loop residues selected!/,
         '                     error message');

    throws_ok { modloop::parse_loop_selection('::10::20:A:31:A:') }
              "saliweb::frontend::InputValidationError",
              '                     empty first residue';
    like($@, qr/No loop residues selected!/,
         '                     error message');
}

# Check read_pdb_file
{
    my $pdb = new File::Temp();
    for my $chain (' ', 'A') {
        for my $resid (1 ... 10) {
            printf $pdb "ATOM      1  CA  ALA %1s%4d      " .
                        "18.511  -1.416  15.632  1.00  6.84           C\n",
                        $chain, $resid;

        }
    }

    seek($pdb, 0, 0);
    my $contents = modloop::read_pdb_file($pdb, 2, ['1', '1'], [' ', 'A'],
                                          ['5', '5'], [' ', 'A']);
    like($contents,
         '/^ATOM\s+1\s+CA\s+ALA     1.*ATOM\s+1\s+CA\s+ALA A  10/ms',
         'read_pdb_file successful read, contents');

    seek($pdb, 0, 0);
    throws_ok { modloop::read_pdb_file($pdb, 2, ['1', '1'], [' ', 'A'],
                                       ['5', '15'], [' ', 'A']) }
              saliweb::frontend::InputValidationError,
              '              missing residue ID';
    like($@, qr/not found in ATOM records in the PDB file: 15:A\. Check/,
         '                                  error message');

    seek($pdb, 0, 0);
    throws_ok { modloop::read_pdb_file($pdb, 2, ['1', '1'], [' ', 'A'],
                                       ['5', '5'], [' ', 'B']) }
              saliweb::frontend::InputValidationError,
              '              missing chain ID';
    like($@, qr/not found in ATOM records in the PDB file: 5:B\. Check/,
         '                                  error message');

    throws_ok { modloop::read_pdb_file(undef, 2, ['1', '1'], [' ', 'A'],
                                       ['5', '5'], [' ', 'A']) }
              saliweb::frontend::InputValidationError,
              '              missing PDB file';
    like($@, qr/not found in ATOM records in the PDB file: 1:,.*Check/,
         '                                  error message');
}

File: frontend/modloop/submit.py
from flask import request
import saliweb.frontend
import re
import itertools
import ihm.format


def handle_new_job():
    user_pdb_name = request.files.get("pdb")
    user_name = request.form.get("name", "")
    email = request.form.get("email")
    modkey = request.form.get("modkey")
    loops = request.form.get("loops")

    saliweb.frontend.check_email(email, required=False)
    saliweb.frontend.check_modeller_key(modkey)
    check_loop_selection(loops)
    check_pdb_name(user_pdb_name)

    (loops, start_res, start_id, end_res,
     end_id, loop_data) = parse_loop_selection(loops)

    # read coordinates from file, and check loop residues
    user_pdb, pdbext = read_pdb_file(user_pdb_name, loops, start_res,
                                     start_id, end_res, end_id)

    job = {}

    job = saliweb.frontend.IncomingJob(user_name)

    # Write PDB/mmCIF input
    with open(job.get_path('input%s' % pdbext), 'wb') as fh:
        fh.writelines(user_pdb)

    # Write loop selection
    with open(job.get_path('loops.tsv'), 'w') as fh:
        fh.write("\t".join(loop_data) + "\n")

    job.submit(email)

    # Pop up an exit page
    loopout = " ".join("%d:%s-%d:%s" % (sr, si, er, ei)
                       for (sr, si, er, ei) in zip(start_res, start_id,
                                                   end_res, end_id))

    return saliweb.frontend.render_submit_template(
        'submit.html', loopout=loopout, user_name=user_name,
        email=email, job=job)


def check_loop_selection(loops):
    """Check for a loop selection"""
    if not loops:
        raise saliweb.frontend.InputValidationError(
            "No loop segments were specified!")


def check_pdb_name(pdb_name):
    """Check if a PDB name was specified"""
    if not pdb_name:
        raise saliweb.frontend.InputValidationError(
            "No coordinate file has been submitted!")


def parse_loop_selection(loops):
    """Split out loop selection and check it"""

    # capitalize and remove spaces
    loops = re.sub(r'\s+', '', loops.upper())
    # replace null chain IDs with a single space
    loops = loops.replace("::", ": :")

    loop_data = loops.split(":")[:-1]

    # Make sure correct number of colons were given
    if len(loop_data) % 4 != 0:
        raise saliweb.frontend.InputValidationError(
            "Syntax error in loop selection: check to make sure you "
            "have colons in the correct place (there should be a "
            "multiple of 4 colons)")

    total_res = 0
    start_res = []
    start_id = []
    end_res = []
    end_id = []
    loops = 0
    while loops*4+3 < len(loop_data) and loop_data[loops*4] != "":
        try:
            start_res.append(int(loop_data[loops*4]))
            end_res.append(int(loop_data[loops*4+2]))
        except ValueError:
            raise saliweb.frontend.InputValidationError(
                "Residue indices are not numeric")
        start_id.append(loop_data[loops*4+1])
        end_id.append(loop_data[loops*4+3])
        # all the selected residues
        total_res += (end_res[-1] - start_res[-1] + 1)

        ################################
        # too long loops rejected
        if ((end_res[-1] - start_res[-1]) > 20
                or start_id[-1] != end_id[-1]
                or (end_res[-1] - start_res[-1]) < 0):
            raise saliweb.frontend.InputValidationError(
                "The loop selected is too long (>20 residues) or "
                "shorter than 1 residue or not selected properly "
                "(syntax problem?) "
                "starting position %d:%s, ending position: %d:%s"
                % (start_res[-1], start_id[-1], end_res[-1], end_id[-1]))
        loops += 1

    ################################
    # too many or no residues rejected
    if total_res > 20:
        raise saliweb.frontend.InputValidationError(
            "Too many loop residues have been selected "
            " (selected: %d > limit:20)!" % total_res)
    if total_res <= 0:
        raise saliweb.frontend.InputValidationError(
            "No loop residues selected!")

    return loops, start_res, start_id, end_res, end_id, loop_data


class FileChecker:
    def __init__(self, fh):
        self.fh = fh


class PdbFileChecker(FileChecker):
    filetype = 'PDB'
    extension = '.pdb'

    def check_residues(self, residues):
        def make_residue_id(chain, residue):
            return "%s:%s" % (str(residue).replace(' ', ''),
                              chain.replace(' ', ''))

        file_contents = self.fh.readlines()
        atom_re = re.compile(b'ATOM.................(.)(....)')
        for line in file_contents:
            m = atom_re.match(line)
            if m:
                chain, res = m.group(1), m.group(2)
                chain = chain.decode('ascii')
                res = res.decode('ascii')
                residues.discard(make_residue_id(chain, res))
        return file_contents


class AtomSiteHandler:
    not_in_file = omitted = unknown = None

    def __init__(self, residues):
        self.residues = residues

    def __call__(self, group_pdb, label_asym_id, label_seq_id, auth_seq_id):
        if group_pdb == 'ATOM':
            seq_id = auth_seq_id if auth_seq_id is not None else label_seq_id
            self.residues.discard("%s:%s" % (seq_id, label_asym_id or ''))


class CifFileChecker(FileChecker):
    filetype = 'mmCIF'
    extension = '.cif'

    def check_residues(self, residues):
        ash = AtomSiteHandler(residues)
        c = ihm.format.CifReader(self.fh, category_handler={'_atom_site': ash})
        try:
            c.read_file()  # read first block
        except ihm.format.CifParserError as err:
            raise saliweb.frontend.InputValidationError(
                "Invalid mmCIF file uploaded: %s" % str(err))
        self.fh.seek(0)
        return self.fh.readlines()


def read_pdb_file(pdb, loops, start_res, start_id, end_res, end_id):
    """Read in uploaded PDB/mmCIF file, and check loop residues"""
    def make_residue_id(chain, residue):
        return "%s:%s" % (str(residue).replace(' ', ''),
                          chain.replace(' ', ''))

    # start/end loop residues
    residues = set(make_residue_id(chain_id, res) for (chain_id, res)
                   in itertools.chain(zip(start_id, start_res),
                                      zip(end_id, end_res)))

    if pdb.filename.lower().endswith('.cif'):
        check = CifFileChecker(pdb)
    else:
        check = PdbFileChecker(pdb)
    file_contents = check.check_residues(residues)

    if len(residues) > 0:
        raise saliweb.frontend.InputValidationError(
            "The following residues were not found in ATOM records in"
            " the %s file: " % check.filetype + ", ".join(sorted(residues)) +
            ". Check that you specified the loop segments correctly, and"
            " that you uploaded the correct %s file." % check.filetype)
    return file_contents, check.extension
